### ISSUE TITLE
Get preg_match warning when using @SuppressWarnings(PHPMD.TooManyFields)

### ISSUE THREAD

**Issue by sitraka-hq** [phpmd/phpmd#808]
- PHPMD version: 2.8.1
- PHP Version: 7.4.7
- Installation type: phar file
- Operating System / Distribution & Version: Debian GNU/Linux 10 (buster)

### Current Behavior
When I use @SuppressWarnings(PHPMD.TooManyFields) to suppress warnings in class level I get the following warning.
```
Warning: preg_match(): Unknown modifier 'd' in phar:///usr/local/bin/phpmd/src/main/php/PHPMD/Node/Annotation.php on line 82
```

### Expected Behavior
I Expect @SuppressWarnings(PHPMD.TooManyFields) to suppress phpmd error without warnings.

### Steps To Reproduce:
My phpmd config looks like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="Ruleset"
         xmlns="http://pmd.sf.net/ruleset/1.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0 http://pmd.sf.net/ruleset_xml_schema.xsd">
    <description>Ruleset for PHP Mess Detector that enforces coding standards</description>
    <rule ref="rulesets/cleancode.xml"/>
    <rule ref="rulesets/codesize.xml"/>
    <rule ref="rulesets/controversial.xml"/>
    <rule ref="rulesets/design.xml">
        <exclude name="CouplingBetweenObjects"></exclude>
    </rule>
    <rule ref="rulesets/naming.xml">
        <exclude name="ShortVariable"/>
    </rule>
    <rule ref="rulesets/unusedcode.xml">
        <!-- PHPMD cannot recognize parameters that are enforced by an interface -->
        <exclude name="UnusedFormalParameter"/>
    </rule>

    <rule ref="rulesets/naming.xml/ShortVariable"
          since="0.2"
          message="Avoid variables with short names like {0}. Configured minimum length is {1}."
          class="PHPMD\Rule\Naming\ShortVariable"
          externalInfoUrl="http://phpmd.org/rules/naming.html#shortvariable">
        <priority>3</priority>
        <properties>
            <property name="minimum" description="Minimum length for a variable, property or parameter name" value="3"/>
            <property name="exceptions" value="id,io" />
        </properties>
    </rule>
    <rule name="rulesets/design.xml/CouplingBetweenObjects"
          since="1.1.0"
          message="The class {0} has a coupling between objects value of {1}. Consider to reduce the number of dependencies under {2}."
          class="PHPMD\Rule\Design\CouplingBetweenObjects"
          externalInfoUrl="https://phpmd.org/rules/design.html#couplingbetweenobjects">
        <description>
            <![CDATA[
A class with too many dependencies has negative impacts on several quality
aspects of a class. This includes quality criteria like stability,
maintainability and understandability
            ]]>
        </description>
        <priority>2</priority>
        <properties>
            <property name="maximum" value="20" description="Maximum number of acceptable dependencies." />
        </properties>
    </rule>
</ruleset>
```
I use the following commands to run it:
```bash
phpmd example.php text phpmd.xml
```

And here an example of php class that reproduce the issue:
```php
<?php
/**
 * File: example.php
 * @SuppressWarnings(PHPMD.TooManyFields)
 */
class MyClass
{
    private $field1;

    private string $field2;

    private $field3;

    private $field4 = [];

    private ?DateTimeImmutable $field5;

    private float $field6;

    private $field7;

    private string $field8;

    private string $field9;

    private $field10;

    private $field11;

    private $field12;

    private $field13;

    private $field14;

    private $field15;

    private $field16;

}
```
## Checks before submitting
* [x] Be sure that there isn't already an issue about this. See: [Issues list](https://github.com/phpmd/phpmd/issues)
* [x] Be sure that there isn't already a pull request about this. See: [Pull requests](https://github.com/phpmd/phpmd/pulls)
* [x] I have added every step to reproduce the bug.
* [x] If possible I added relevant code examples.
* [x] This issue is about 1 bug and nothing more.
* [x] The issue has a descriptive title. For example:  "JSON rendering failed on Windows for filenames with space".

**Comment by MarkVaughn** [comment]
Link to the code with the issue
https://github.com/phpmd/phpmd/blob/2.8.1/src/main/php/PHPMD/Node/Annotation.php#L82
The error sounds like the regex isn't being escaped right. 
Is there any code snippet you can create to run which triggers the error `phpmd example.php text phpmd.xml` ?

**Comment by kylekatarnls** [comment]
As a first sigh, `->getName()` should not contains RegExp characters. Rules names should be alphanumeric.

As the message error si `d` as a modifier, it means it found a `/` before `d` and as `l` is not a modifier neither, I suspect, you may have written `/` instead of `l` (lower `L`) in `TooManyFields`.

Can you try to remove and copy-paste this exact string instead? `@SuppressWarnings(PHPMD.TooManyFields)`

**Comment by sitraka-hq** [comment]
> Is there any code snippet you can create to run which triggers the error `phpmd example.php text phpmd.xml` ?

I just added that code in the issue description.



> As a first sigh, `->getName()` should not contains RegExp characters. Rules names should be alphanumeric.

> Can you try to remove and copy-paste this exact string instead? `@SuppressWarnings(PHPMD.TooManyFields)`

I tried and got the same error.

**Comment by kylekatarnls** [comment]
The `->getName()` actually contains `rulesets/design.xml/CouplingBetweenObjects`

Now I have to find if `@SuppressWarnings()` should support that or if the config should be done differently.

**Comment by kylekatarnls** [comment]
I checked the doc, the usage is fair, so I will fix the annotation to ignore the path of the re-defined rules.
